rfc/accepted/0021-stream-processors.md

```diff
@@ -0,0 +1,205 @@
+                  ┌─────────────────────────┐
+                  │                         ├──package data─▶
+───package data──▶│framing package processor│
+                  │                         ├──stream data──▶
+                  └─────────────────────────┘
+```
+
+Framing packet processors take package data and return new package data, but with framing information, their output can either be treated as streaming or packet data. These packet processors are required for transitioning from one model to another, but they only work in the outbound direction—they have an analougus 'framing' stream processor on the inbound side. An example here would be the `separate` processor delemiting evens by newlines or the `length-prefix` processor.
+
+### stream processors
+
+Stream processors have no notion of events. They operate on a stream of bytes and produce a transformed stream of bytes.
+
+#### 'pure' stream processors
+
+```
+                   ┌────────────────────────┐
+────stream data───▶│ pure stream processor  │────stream data───▶
+                   └────────────────────────┘
+                   ┌────────────────────────┐
+◀───stream data────│ pure stream processor  │◀───stream data────
+                   └────────────────────────┘
+```
+'pure' stream processors take and produce stream data. Any boundaries between event data in the stream are ignored and likely change as data is passed through the processor. An example would be streaming zx2 compression of events into an S3 object.
+
+
+#### 'framing' stream processors
+
+```
+                   ┌─────────────────────────┐
+◀───package data───┤                         │
+                   │framing stream processor │◀──stream data───
+◀──streaming data──┤                         │
+                   └─────────────────────────┘
+```
+
+Framing stream processors take streaming data and subdivide it into framed package data. The result can be treated as packet data or subdivided as stream data. This stream processor is required to transition from one model to another. Still, they only work in the inbound direction - they have an analogous 'framing' packet processor on the outbound side. An example here would be the `separate` processor splitting a stream into evens delimited by newlines, or the `length-prefix` processor that reads the length of the framed event and extracts as many bytes.
+
+### changes to codecs
+
+To enable codecs to produce hygenic errors and avoid misconfiguration codecs need to be able to declare if they produce or accept stream or package data. Any codecs, impemented as of the writing of this RFC, purely act on packet data.
+
+### changes to connectors
+
+Just  as codecs, connectors need to be able to declare if they work with stream or package data, below a survey of a supset of connectors implemented today. As mentioned above, structured connectors are not affected.
+
+| Connector |  type  |
+|-----------|--------|
+| tcp       | stream |
+| udp       | packet |
```

Comment:
Why can't the UDP processor produce and handle a stream. The handling should work, shouldn't it? While it is logically a stream, we'd still receive data-chunks of indefinite size we could put in UDP packets. While this might be dangerous, it is possible, no?

Reply:
There are two main reasons:

First, UDP and other packet-based transports require packages. You need to chunk the data into some bite-sized portions for it. Unlike TCP, you can't just can't just continuously stream data into an UDP socket. This could be worked around by adding a chunking processor that takes a stream and cuts it into package - but this would be back to a packet based  delivery.

The second reason is more damning, that is that streams need to be contious. In other words data needs to arrive and arrive in order. UDP can't do that.

To elaborate, let's look at four realistic streaming processors:

## separate
We concatinate events by separating each packet with some kind of delimiter. The first event starts at the beginning, after that a new one starts every delimiter.

UDP won't work here because:  If packets arrive out of order or go missing the data can't be reassembled faithfully. - this will lead to intermediate data corruption but will self correct eventually

## length prefix

We prefix each encoded event with it's length to know it's boundaries.

UDP won't work here because:  If packets arrive out of order or go missing we likely never align on a proper prefix again and get a garbled mess - this will likely lead to complete data corruption

## compression

The data stream is compressed as an entire "file":

UDP won't work here because:  If packets arrive out of order or go missing we get an invalid "file" and can't decompress it any more - this will lead to total dataloss

## encryptiption

The data stream is encrypted as an entire "file":

UDP won't work here because:  If packets arrive out of order or go missing we get an invalid "file" and can't decrypted it any more - this will lead to total dataloss